src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/StockInMemoryRepositoryImpl.kt

```diff
@@ -1,52 +1,100 @@
 package systems.ajax.malov.stockanalyzer.repository.impl
 
 import org.springframework.stereotype.Repository
+import systems.ajax.malov.stockanalyzer.constant.NUMBER_OF_HISTORY_RECORDS_PER_STOCK
+import systems.ajax.malov.stockanalyzer.constant.WEIGHTED_COEFFICIENT_FOR_PRICE_COEFFICIENTS
 import systems.ajax.malov.stockanalyzer.entity.Stock
 import systems.ajax.malov.stockanalyzer.repository.StockRepository
 import java.math.BigDecimal
 import java.time.Duration
 import java.time.Instant
 import java.util.*
+import kotlin.reflect.KProperty1
 
 
 @Repository
 class StockInMemoryRepositoryImpl : StockRepository {
     private val db: HashMap<UUID, Stock> = HashMap()
 
-    override fun insertAll(stocks: List<Stock>): MutableCollection<Stock> {
+    override fun insertAll(stocks: List<Stock>): List<Stock> {
         stocks.forEach {
             val id = UUID.randomUUID()
             it.id = id
             db[id] = it
         }
-        return stocks.toMutableList()
+        return stocks
     }
 
-    override fun findFiveBestStocksToBuy(): List<Pair<String?, List<Stock>>> {
+    override fun findTopNStockSymbolsWithStockData(n: Int): List<Pair<String, List<Stock>>> {
         if (db.isEmpty()) return emptyList()
 
-        val maxPercentChange = db.asSequence().mapNotNull { it.value.percentChange }.maxBy { it }
-        val maxChange = db.asSequence().mapNotNull { it.value.change }.maxBy { it }
-
-        return db.asSequence().filterNot { it.value.symbol == null }.filter { stock ->
-            stock.value.dateOfRetrieval?.isBefore(Instant.now().plus(Duration.ofHours(1))) == true
-        }.map { it.value }.groupBy { it.symbol }.toList().sortedByDescending { (_, stockList) ->
-            val avgChange = stockList.mapNotNull { it.change }
-                .fold(BigDecimal.ZERO) { acc, bigDecimal -> acc + bigDecimal } / BigDecimal.valueOf(stockList.size.toLong())
-            val avgPercentChange = stockList.mapNotNull { it.percentChange }
-                .fold(BigDecimal.ZERO) { acc, bigDecimal -> acc + bigDecimal } / BigDecimal.valueOf(stockList.size.toLong())
-
-            ((avgChange / maxChange) * BigDecimal.valueOf(0.5)) + ((avgPercentChange / maxPercentChange) * BigDecimal.valueOf(
-                0.5
-            ))
-        }.take(5)
+        val maxPercentChange = getMaxBigDecimal(Stock::percentChange)
+        val maxChange = getMaxBigDecimal(Stock::change)
+
+        val timeOfRequest = Instant.now()
+
+        return db.asSequence()
```

Comment:
```suggestion
        return db.values.asSequence()
```
I don't see usage of key in the chain